File: python/year_2016/day_06.py
"""
https://adventofcode.com/2016/day/6
"""

from collections import Counter
from pathlib import Path
from statistics import mode


def part_1(text):
    return "".join(map(mode, zip(*text.splitlines())))


def part_2(text):
    def antimode(message):
        counts = Counter(message)
        return min(counts, key=counts.get)

    return "".join(map(antimode, zip(*text.splitlines())))


def _read_input() -> str:
    return (Path(__file__).with_name("test_data") / "day_06.txt").read_text()


EXAMPLE = """eedadn
drvtee
eandsr
raavrd
atevrs
tsrnev
sdttsa
rasrtv
nssdts
ntnada
svetve
tesnvt
vntsnd
vrdear
dvrsen
enarar
"""


def test_part_1_input():
    assert part_1(_read_input()) == "gyvwpxaz"


def test_part_1_example():
    assert part_1(EXAMPLE) == "easter"


def test_part_2_input():
    assert part_2(_read_input()) == "jucfoary"


def test_part_2_example():
    assert part_2(EXAMPLE) == "advent"

File: rust/year_2016/src/day_06.rs
pub fn part_1(input: &str) -> String {
    counts(input)
        .into_iter()
        .map(|counts| {
            let (i, _) = counts
                .into_iter()
                .enumerate()
                .max_by_key(|&(_, c)| c)
                .unwrap();
            char::from(b'a' + u8::try_from(i).unwrap())
        })
        .collect()
}

pub fn part_2(input: &str) -> String {
    counts(input)
        .into_iter()
        .map(|counts| {
            let (i, _) = counts
                .into_iter()
                .enumerate()
                .filter(|&(_, c)| c > 0)
                .min_by_key(|&(_, c)| c)
                .unwrap();
            char::from(b'a' + u8::try_from(i).unwrap())
        })
        .collect()
}

pub fn counts(input: &str) -> Vec<[usize; 26]> {
    let n = input.find('\n').unwrap_or(input.len());
    input.lines().fold(vec![[0; 26]; n], |mut counts, line| {
        counts.iter_mut().zip(line.bytes()).for_each(|(c, b)| {
            let i: usize = (b - b'a').into();
            c[i] += 1;
        });
        counts
    })
}

#[cfg(test)]
mod test {
    use test_case::test_case;

    const INPUT: &str = include_str!("../test_data/day_06.txt");

    const EXAMPLE: &str = "eedadn
drvtee
eandsr
raavrd
atevrs
tsrnev
sdttsa
rasrtv
nssdts
ntnada
svetve
tesnvt
vntsnd
vrdear
dvrsen
enarar";

    #[test_case(EXAMPLE => "easter")]
    #[test_case(INPUT => "gyvwpxaz")]
    fn part_1(input: &str) -> String {
        super::part_1(input)
    }

    #[test_case(EXAMPLE => "advent")]
    #[test_case(INPUT => "jucfoary")]
    fn part_2(input: &str) -> String {
        super::part_2(input)
    }
}
